runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/JsonElementSerializer.kt

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2017-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.json.internal
+
+import kotlinx.serialization.*
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.json.*
+
+@Serializer(forClass = JsonElement::class)
+internal object JsonElementSerializer : KSerializer<JsonElement> {
+    override val descriptor: SerialDescriptor = object : SerialClassDescImpl("JsonElementSerializer") {
+        override val kind: SerialKind
+            get() = UnionKind.SEALED
+
+        init {
+            addElement("JsonElement")
+        }
+    }
+
+    override fun serialize(encoder: Encoder, obj: JsonElement) {
+        when (obj) {
+            is JsonPrimitive -> JsonPrimitiveSerializer.serialize(encoder, obj)
+            is JsonObject -> JsonObjectSerializer.serialize(encoder, obj)
+            is JsonArray -> JsonArraySerializer.serialize(encoder, obj)
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): JsonElement {
+        val input = decoder as? JsonInput ?: error("JsonElement is serializable only when used by Json")
```

Comment:
maybe better correct this error message to 'deserializable only'? Because it can be serialized